lib/Onyx.js

```diff
@@ -551,34 +594,31 @@ function mergeCollection(collectionKey, collection) {
         }
     });
 
-    const existingKeyCollection = {};
-    const newCollection = {};
     return getAllKeys()
-        .then((keys) => {
-            _.each(collection, (data, dataKey) => {
-                if (keys.includes(dataKey)) {
-                    existingKeyCollection[dataKey] = data;
-                } else {
-                    newCollection[dataKey] = data;
-                }
-            });
-
-            const keyValuePairsForExistingCollection = _.reduce(existingKeyCollection, (finalArray, val, key) => ([
-                ...finalArray,
-                [key, JSON.stringify(val)],
-            ]), []);
-            const keyValuePairsForNewCollection = _.reduce(newCollection, (finalArray, val, key) => ([
-                ...finalArray,
-                [key, JSON.stringify(val)],
-            ]), []);
+        .then((persistedKeys) => {
+            // Split to keys that exist in storage and keys that don't
+            const [existingKeys, newKeys] = _.chain(collection)
+                .keys()
+                .partition(key => persistedKeys.includes(key))
+                .value();
+
+            const existingKeyCollection = _.pick(collection, existingKeys);
+            const newCollection = _.pick(collection, newKeys);
+            const keyValuePairsForExistingCollection = prepareKeyValuePairsForStorage(existingKeyCollection);
+            const keyValuePairsForNewCollection = prepareKeyValuePairsForStorage(newCollection);
 
             // New keys will be added via multiSet while existing keys will be updated using multiMerge
             // This is because setting a key that doesn't exist yet with multiMerge will throw errors
             const existingCollectionPromise = AsyncStorage.multiMerge(keyValuePairsForExistingCollection);
             const newCollectionPromise = AsyncStorage.multiSet(keyValuePairsForNewCollection);
 
+            // Capture (non-stringified) changes to cache
+            cache.merge(_.pairs(collection));
+
+            // Optimistically inform collection subscribers
+            keysChanged(collectionKey, collection);
+
             return Promise.all([existingCollectionPromise, newCollectionPromise])
-                .then(() => keysChanged(collectionKey, collection))
                 .catch(error => evictStorageAndRetry(error, mergeCollection, collection));
```

Comment:
Shouldn't we also call `keysChanged` for the individual collection items

E.g. collection is just the collection name 
![image](https://user-images.githubusercontent.com/12156624/119580473-23b3e380-bdc9-11eb-87eb-6c4a6f99fd29.png)

It seems `mergeCollection` was called to merge new content to all these collections but no subscriber would be notified for that
![image](https://user-images.githubusercontent.com/12156624/119580697-94f39680-bdc9-11eb-9650-03a1a0b71638.png)

Isn't it possible to subscribe to a particular collection item like `report_71402263` ?
I think currently it doesn't matter because when a collection is updated we're usually already have the updated data rendered, but this might not always be the case

Reply:
> > Shouldn't we also call keysChanged for the individual collection items
> > E.g. collectionKey is just the collection name
> 
> Correct, but we loop over all of the possible subscribers to find which ones need to be updated.

I see that now, I didn't know that `keysChanged` does that internally